lib/taste_tester/commands.rb

```diff
@@ -210,5 +211,187 @@ def self.upload
+      repo = BetweenMeals::Repo.get(
+        TasteTester::Config.repo_type,
+        TasteTester::Config.repo,
+        logger,
+      )
+      if repo && !repo.exists?
+        fail "Could not open repo from #{TasteTester::Config.repo}"
+      end
+
+      # Compute the set of changes using Between Meals or a custom hook
+      if TasteTester::Config.use_custom_changeset_hook
+        changeset = TasteTester::Hooks.custom_changeset(repo)
+      else
+        changeset = _find_changeset(repo)
+      end
+
+      # Use the changeset computed earlier to find modified roles
+      if TasteTester::Config.use_custom_impact_hook
+        impact_roles = TasteTester::Hooks.custom_impact(changeset)
+      else
+        impact_roles = _find_impact(changeset)
+      end
+
+      if impact_roles.empty?
+        logger.warn('No impacted roles were found.')
+      else
+        logger.warn('The following roles have modified dependencies. ' +
+                    'Please test a host in each of these roles.')
+        impact_roles.each { |r| logger.warn("\t#{r}") }
+      end
+
+      # Do any post processing required on the list of impacted roles
+      unless TasteTester::Config.skip_post_impact_hook
+        TasteTester::Hooks.post_impact(impact_roles)
+      end
+    end
+
+    def self._find_changeset(repo)
+      # We want to compare changes in the current directory (working set)
+      # with the "most recent" commit in the VCS. For SVN, this will be the
+      # latest commit on the checked out repository (i.e. 'trunk'). Git/Hg
+      # may have different tags or labels assigned to the "master" branch,
+      # (i.e. 'master', 'stable', etc.) and should be configured if different
+      # than the defaults.
+      start_ref = ''
+      type = repo.class.name.split('::').last
+      case type
+      when 'Svn'
+        start_ref = repo.latest_revision
+      when 'Git'
+        start_ref = TasteTester::Config.vcs_master_git
+      when 'Hg'
+        start_ref = TasteTester::Config.vcs_master_hg
+      end
+
+      changeset = BetweenMeals::Changeset.new(
+        logger,
+        repo,
+        start_ref,
+        nil, # compare with current working copy
+        {
+          :cookbook_dirs =>
+            TasteTester::Config.relative_cookbook_dirs,
+          :role_dir =>
+            TasteTester::Config.relative_role_dir,
+        },
+        @track_symlinks,
+      )
+
+      return changeset
+    end
+
+    def self._find_impact(changeset)
+      if TasteTester::Config.relative_cookbook_dirs.length > 1
+        logger.error('Knife deps does not support multiple cookbook paths.')
+        logger.error('Flatten the repository or use the custom_impact hook.')
+        exit(1)
+      end
+
+      cookbooks = Set.new(changeset.cookbooks)
+      roles = Set.new(changeset.roles)
+
+      if cookbooks.empty? && roles.empty?
+        logger.error('No cookbooks or roles have been modified.')
+        exit(1)
```

Comment:
Yea, just return here.

Reply:
Agreed, I will change this to return an empty set.

> Oh one other thought I had... TT/BM/GD supports databags. FB Prod doesn't use them but FB CPE does, and so do non-FB users... I think it would make sense for `impact` to print out the databags changing.
> 
> And... also, `impact` probably should have a `--json` option for machine-parsability for aftertaste...

Good point. I will look at adding databag support and a json option

Reply:
Agreed. I will change this to return an empty set.